### DISCUSSION TITLE
Ben Eater 6502 breadboard computer support?

### DISCUSSION THREAD

**Discussion by mikeakers** [cc65/cc65#2740] [General]
I'm working through building a [Ben Eater 6502 computer kit](https://eater.net/6502) (aka BE6502) and I got to thinking it could be fun to try to write programs for it in C. This lead me to find this project and I'm wondering if anyone's tried targeting this computer with this compiler? If not I may try it myself, can anyone point me to a good place to get started? I'm thinking the current Atari 2600 support may be a good place to get ideas from, since the 2600 is such a simple machine.

**Comment by polluks2** [comment]
Please take a look at https://cc65.github.io/doc/customizing.html.

**Comment by mrdudz** [comment]
I'd start with the sim6502 crt0 and linker script, and use -none target.

**Comment by oliverschmidt** [comment]
I'm pretty surprised. What makes you prefer the `sim6502` files over the `none` files?

**Comment by mrdudz** [comment]
no idea, no particular reason :)

**Comment by oliverschmidt** [comment]
Then let's rather propose the `none` cfg, crt0 and lib ;-)

**Comment by neilbalch** [comment]
Hello, all! I'm curious if anything came out of this discussion. I'm also thinking about using `cc65` for a bare/custom 6502 breadboard computer project, but not sure where to start with making/adapting a custom target.

I see a lot of implementation work around making binaries that run inside the OSes of the commercial computers, but not much detail about making it work without all that infrastructure.

**Comment by mrdudz** [comment]
did you read https://cc65.github.io/doc/customizing.html ? what information are you missing?

**Comment by polluks** [comment]
Well, we have a few targets without OS at all:
NES, Gamate, Supervision... just bang the hardware.

**Comment by neilbalch** [comment]
> 
> 
> did you read https://cc65.github.io/doc/customizing.html ? what information are you missing?

I know this isn't a terribly helpful response, but I just feel kindof confused. I'm a noob when it comes to 6502 programming and assembly-level stuff, so I've been having a bit of a hard time following the linked page's instructions for what specific address space partitions to use where and in what files to put the code segments.

**Comment by groessler** [comment]
I've clicked on the link to Beneater in the first post. Look like an interesting thing, I haven't watched the videos.

I didn't find documentation about which devices are at what location in the memory map. And no memory map at all. Such information is crucial if you want to target cc65 to this device.

**Comment by neilbalch** [comment]
> 
> 
> I've clicked on the link to Beneater in the first post. Look like an interesting thing, I haven't watched the videos.
> 
> I didn't find documentation about which devices are at what location in the memory map. And no memory map at all. Such information is crucial if you want to target cc65 to this device.

Well, for the record, the rest of the videos in the series discuss the address space layout. But, I'm asking more generally about how to set up a target for a generic 6502 computer with a "custom" memory map. I'm asking _(at least I think I am)_ about the procedure for determining what to tell `cc65` and how, not the exact numbers to use.

**Comment by groessler** [comment]
You should read and understand ld65's config files: https://cc65.github.io/doc/ld65.html#s5  This aspect appears to me to be the most complicated part for people. In the config file you specify which part of the program "goes" into which memory area. But, spoiler, these parts don't automatically get loaded there, it's just how the linker resolves address references.

On the Ben Eater programs appear to be stored in EPROMS. So the program is in read-only memory. Look at some "video console" targets like creativision.cfg, nes.cfg, gamate.cfg, pce.cfg, supervision.cfg how it's done there. The data (which is potentially written to during program execution) is stored in the ROM, but copied to RAM at startup of the program in crt0.s. This is reflected in the linker config files with "load = XXX, run = RAM". The "run =" instructs the linker to treat this segment as being located at RAM.

If you can show me the memory map, I could look at your linker config file if it looks good.

**Comment by neilbalch** [comment]
> 
> 
> You should read and understand ld65's config files: https://cc65.github.io/doc/ld65.html#s5 This aspect appears to me to be the most complicated part for people. In the config file you specify which part of the program "goes" into which memory area. But, spoiler, these parts don't automatically get loaded there, it's just how the linker resolves address references.
> 
> On the Ben Eater programs appear to be stored in EPROMS. So the program is in read-only memory. Look at some "video console" targets like creativision.cfg, nes.cfg, gamate.cfg, pce.cfg, supervision.cfg how it's done there. The data (which is potentially written to during program execution) is stored in the ROM, but copied to RAM at startup of the program in crt0.s. This is reflected in the linker config files with "load = XXX, run = RAM". The "run =" instructs the linker to treat this segment as being located at RAM.
> 
> If you can show me the memory map, I could look at your linker config file if it looks good.

Thanks, I'd really appreciate that! I've been busy with class so far this week, but I'll try to see what I can get done myself soon.

**Comment by incyi** [comment]
Any progress on this topic ?

**Comment by neilbalch** [comment]
> Any progress on this topic?

Yes, my friend and I worked this into a functional project:

- [Demo workspace](https://github.com/neilbalch/cc65-customizing-demo)
- [Project workspace](https://github.com/ucsbieee/mapache64)
